File: src/python/xerxes/runtime/distribution.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Distribution and packaging helpers for ``xerxes`` install/release tooling.

Detects whether the current host is Termux/Android or WSL2, renders a ready-to-commit
Homebrew formula, strips extras that lack Android wheels, and provides shell
and PowerShell install snippets used by the docs and release scripts.
"""

from __future__ import annotations

import os
import platform
import sys
from dataclasses import dataclass


@dataclass
class PlatformInfo:
    """Snapshot of the running host used for distribution decisions.

    Attributes:
        system: ``platform.system()`` value (``"Darwin"``, ``"Linux"``,
            ``"Windows"``).
        release: Kernel/release string as returned by ``platform.release()``.
        is_termux: ``True`` when Xerxes is running under Termux/Android.
        is_wsl: ``True`` when the host is Windows Subsystem for Linux.
        python: Python version (``"3.11.7"``-style).
    """

    system: str
    release: str
    is_termux: bool
    is_wsl: bool
    python: str

    @property
    def is_darwin(self) -> bool:
        """``True`` when running on macOS."""
        return self.system == "Darwin"

    @property
    def is_linux(self) -> bool:
        """``True`` when running on Linux (including WSL/Termux)."""
        return self.system == "Linux"

    @property
    def is_windows(self) -> bool:
        """``True`` when running on native Windows."""
        return self.system == "Windows"


def detect_platform() -> PlatformInfo:
    """Return ``PlatformInfo`` describing the current host."""
    sysname = platform.system()
    release = platform.release()
    is_termux = bool(os.environ.get("PREFIX", "").startswith("/data/data/com.termux/")) or "termux" in release.lower()
    is_wsl = "microsoft" in release.lower() or "WSL_DISTRO_NAME" in os.environ
    return PlatformInfo(
        system=sysname,
        release=release,
        is_termux=is_termux,
        is_wsl=is_wsl,
        python=sys.version.split()[0],
    )


HOMEBREW_FORMULA_TEMPLATE = """\
class XerxesAgent < Formula
  include Language::Python::Virtualenv

  desc "Multi-agent orchestration framework + terminal coding agent"
  homepage "{homepage}"
  url "{tarball_url}"
  sha256 "{sha256}"
  license "Apache-2.0"

  depends_on "python@3.11"

  def install
    virtualenv_install_with_resources
  end

  test do
    system "#{{bin}}/xerxes", "--help"
  end
end
"""


def render_homebrew_formula(
    *, version: str, tarball_url: str, sha256: str, homepage: str = "https://github.com/erfanzar/Xerxes-Agents"
) -> str:
    """Render a ready-to-commit Homebrew formula for the given release.

    Args:
        version: Released version, embedded into the formula header.
        tarball_url: URL pointing at the source tarball.
        sha256: SHA-256 of the tarball.
        homepage: Project homepage; defaults to the GitHub repo.
    """
    return HOMEBREW_FORMULA_TEMPLATE.format(version=version, tarball_url=tarball_url, sha256=sha256, homepage=homepage)


TERMUX_EXTRA_EXCLUDED = frozenset(
    {
        "faster_whisper",
        "sentence_transformers",
        "playwright",
        "edge_tts",
        "ctranslate2",
        "onnxruntime",
        "elevenlabs",
        "sounddevice",
    }
)


def termux_filter_extras(extras: dict[str, list[str]]) -> dict[str, list[str]]:
    """Strip dependencies that lack Android wheels from a Termux build.

    Args:
        extras: Mapping of extra name → dependency list (e.g. as parsed from
            ``pyproject.toml``).

    Returns:
        A copy of ``extras`` with any dependency matching
        :data:`TERMUX_EXTRA_EXCLUDED` removed. Comparison is case-insensitive
        and ignores ``-``/``_`` so ``faster-whisper`` matches the
        ``faster_whisper`` blocklist entry.
    """

    def _normalize(s: str) -> str:
        """Lower-case ``s`` and collapse ``-`` into ``_`` for matching."""
        return s.lower().replace("-", "_")

    out: dict[str, list[str]] = {}
    for extra_name, deps in extras.items():
        kept = [d for d in deps if not any(forbidden in _normalize(d) for forbidden in TERMUX_EXTRA_EXCLUDED)]
        out[extra_name] = kept
    return out


SHELL_INSTALL_SNIPPET = """\
#!/usr/bin/env bash
set -euo pipefail
if ! command -v uv >/dev/null 2>&1; then
  echo "Installing uv..."
  curl -LsSf https://astral.sh/uv/install.sh | sh
  export PATH="$HOME/.local/bin:$PATH"
fi
uv tool install xerxes-agent
echo "✓ xerxes installed. Run 'xerxes' to start."
"""


POWERSHELL_INSTALL_SNIPPET = """\
$ErrorActionPreference = 'Stop'
if (-not (Get-Command uv -ErrorAction SilentlyContinue)) {
  Write-Host "Installing uv..."
  Invoke-Expression "& $((Invoke-WebRequest -UseBasicParsing https://astral.sh/uv/install.ps1).Content)"
}
uv tool install xerxes-agent
Write-Host "✓ xerxes installed."
"""


__all__ = [
    "HOMEBREW_FORMULA_TEMPLATE",
    "POWERSHELL_INSTALL_SNIPPET",
    "SHELL_INSTALL_SNIPPET",
    "TERMUX_EXTRA_EXCLUDED",
    "PlatformInfo",
    "detect_platform",
    "render_homebrew_formula",
    "termux_filter_extras",
]

File: xerxes/src/runtime/distribution.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { existsSync } from "node:fs";
import { release as osRelease } from "node:os";
import { resolve } from "node:path";

const DEFAULT_HOMEPAGE = "https://github.com/erfanzar/Xerxes-Agents";
const DEFAULT_BUN_FORMULA = "oven-sh/bun/bun";
const DEFAULT_ENTRYPOINT = "xerxes/src/cli.ts";

/** Resolve the OpenTUI artifact in either a source checkout or staged release package. */
export function resolveTuiEntry(
  moduleDirectory: string,
  fileExists: (path: string) => boolean = existsSync,
): string | undefined {
  const candidates = [
    resolve(moduleDirectory, "../ui/entry.js"),
    resolve(moduleDirectory, "ui/entry.js"),
    resolve(moduleDirectory, "../dist/ui/entry.js"),
  ];
  return candidates.find(fileExists);
}

/** Snapshot of the Bun host used to select distribution guidance. */
export interface PlatformInfo {
  readonly bunVersion: string;
  readonly isDarwin: boolean;
  readonly isLinux: boolean;
  readonly isTermux: boolean;
  readonly isWindows: boolean;
  readonly isWsl: boolean;
  readonly platform: string;
  readonly release: string;
}

/** Injectable process values make platform detection deterministic in callers and tests. */
export interface DetectPlatformOptions {
  readonly bunVersion?: string;
  readonly environment?: Readonly<Record<string, string | undefined>>;
  readonly platform?: string;
  readonly release?: string;
}

/** Detect Bun's host platform, Termux, and WSL without mutating process state. */
export function detectPlatform(
  options: DetectPlatformOptions = {},
): PlatformInfo {
  const platform = options.platform ?? process.platform;
  const release = options.release ?? osRelease();
  const environment = options.environment ?? process.env;
  const normalizedRelease = release.toLowerCase();
  const prefix = environment.PREFIX ?? "";
  const isTermux =
    prefix.startsWith("/data/data/com.termux/") ||
    normalizedRelease.includes("termux") ||
    environment.TERMUX_VERSION !== undefined;
  const isWsl =
    normalizedRelease.includes("microsoft") ||
    environment.WSL_DISTRO_NAME !== undefined;

  return Object.freeze({
    bunVersion: options.bunVersion ?? Bun.version,
    isDarwin: platform === "darwin",
    isLinux: platform === "linux",
    isTermux,
    isWindows: platform === "win32",
    isWsl,
    platform,
    release,
  });
}

/** Packages with native/runtime assumptions that are not supported by the Termux Bun target. */
export const TERMUX_UNSUPPORTED_PACKAGES: ReadonlySet<string> = new Set([
  "@playwright/test",
  "onnxruntime-node",
  "playwright",
  "playwright-core",
]);

/** Return whether an npm package should be omitted from a Termux Bun install. */
export function isTermuxUnsupportedPackage(packageName: string): boolean {
  return TERMUX_UNSUPPORTED_PACKAGES.has(normalizePackageName(packageName));
}

/**
 * Copy a package.json-style dependency mapping while omitting known Termux-incompatible packages.
 *
 * This only filters dependency declarations; it does not make an arbitrary npm package Android-safe.
 */
export function filterTermuxDependencies(
  dependencies: Readonly<Record<string, string>>,
): Readonly<Record<string, string>> {
  const filtered: Record<string, string> = {};
  for (const [packageName, version] of Object.entries(dependencies)) {
    if (!isTermuxUnsupportedPackage(packageName)) {
      filtered[packageName] = version;
    }
  }
  return Object.freeze(filtered);
}

/** Inputs needed to render a Bun-oriented Homebrew formula template. */
export interface HomebrewFormulaOptions {
  readonly bunFormula?: string;
  readonly entrypoint?: string;
  readonly homepage?: string;
  readonly sha256: string;
  readonly tarballUrl: string;
  readonly version: string;
}

/**
 * A source formula template for a Bun release artifact.
 *
 * It intentionally does not assert that a package has been published or that a tap has accepted it.
 */
export const HOMEBREW_BUN_FORMULA_TEMPLATE = `class XerxesAgent < Formula
  desc "Multi-agent orchestration framework + terminal coding agent"
  homepage {homepage}
  url {tarball_url}
  version {version}
  sha256 {sha256}
  license "Apache-2.0"

  depends_on {bun_formula}

  def install
    libexec.install Dir["*"]
    cd libexec do
      system Formula[{bun_formula}].opt_bin/"bun", "install", "--production", "--frozen-lockfile"
    end
    (bin/"xerxes").write <<~EOS
      #!/bin/bash
      exec "#{Formula[{bun_formula}].opt_bin}/bun" "#{libexec}/{entrypoint}" "$@"
    EOS
    chmod 0755, bin/"xerxes"
  end

  test do
    system bin/"xerxes", "--help"
  end
end
`;

/** Render a reviewable Homebrew formula for a Bun source tarball. */
export function renderHomebrewFormula(options: HomebrewFormulaOptions): string {
  const homepage = options.homepage ?? DEFAULT_HOMEPAGE;
  const bunFormula = options.bunFormula ?? DEFAULT_BUN_FORMULA;
  const entrypoint = options.entrypoint ?? DEFAULT_ENTRYPOINT;
  return HOMEBREW_BUN_FORMULA_TEMPLATE.replaceAll(
    "{homepage}",
    rubyString(homepage, "homepage"),
  )
    .replaceAll("{tarball_url}", rubyString(options.tarballUrl, "tarballUrl"))
    .replaceAll("{version}", rubyString(options.version, "version"))
    .replaceAll("{sha256}", rubyString(options.sha256, "sha256"))
    .replaceAll("{bun_formula}", rubyString(bunFormula, "bunFormula"))
    .replaceAll("{entrypoint}", formulaPath(entrypoint, "entrypoint"));
}

/**
 * A Bun install template that deliberately requires a release-provided package or source spec.
 * It does not imply that Xerxes is already published to an npm registry.
 */
export const BUN_SHELL_INSTALL_SNIPPET = `#!/usr/bin/env bash
set -euo pipefail
if ! command -v bun >/dev/null 2>&1; then
  echo "Installing Bun..."
  curl -fsSL https://bun.sh/install | bash
  export BUN_INSTALL="\${BUN_INSTALL:-$HOME/.bun}"
  export PATH="$BUN_INSTALL/bin:$PATH"
fi
: "\${XERXES_PACKAGE:?Set XERXES_PACKAGE to a release-approved npm package or source spec.}"
bun add --global "$XERXES_PACKAGE"
echo "Xerxes installed through Bun. Restart your shell, then run 'xerxes'."
`;

/** PowerShell counterpart to {@link BUN_SHELL_INSTALL_SNIPPET}. */
export const BUN_POWERSHELL_INSTALL_SNIPPET = `$ErrorActionPreference = 'Stop'
if (-not (Get-Command bun -ErrorAction SilentlyContinue)) {
  Write-Host "Installing Bun..."
  powershell -c "irm bun.sh/install.ps1|iex"
}
if (-not $env:XERXES_PACKAGE) {
  throw "Set XERXES_PACKAGE to a release-approved npm package or source spec."
}
bun add --global $env:XERXES_PACKAGE
Write-Host "Xerxes installed through Bun."
`;

function normalizePackageName(packageName: string): string {
  return packageName.trim().toLowerCase().replaceAll("_", "-");
}

function rubyString(value: string, field: string): string {
  if (!value.trim()) {
    throw new TypeError(`${field} must not be empty`);
  }
  if (value.includes("\r") || value.includes("\n")) {
    throw new TypeError(`${field} must not contain a newline`);
  }
  return JSON.stringify(value);
}

function formulaPath(value: string, field: string): string {
  if (!value.trim()) {
    throw new TypeError(`${field} must not be empty`);
  }
  if (
    value.startsWith("/") ||
    value.includes("\\") ||
    value.includes("\r") ||
    value.includes("\n") ||
    value.split("/").some((segment) => segment === "..")
  ) {
    throw new TypeError(`${field} must be a relative, newline-free path`);
  }
  return value;
}
